tmk_core/common/keyboard.c

```diff
@@ -64,20 +64,41 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 
 #ifdef MATRIX_HAS_GHOST
+static uint16_t matrix_ghost_check[MATRIX_ROWS];
 static bool has_ghost_in_row(uint8_t row)
 {
-    matrix_row_t matrix_row = matrix_get_row(row);
-    // No ghost exists when less than 2 keys are down on the row
-    if (((matrix_row - 1) & matrix_row) == 0)
+    matrix_row_t matrix_row = (matrix_get_row(row) & matrix_ghost_check[row]);
+    /* No ghost exists when less than 2 keys are down on the row.
+    If there are "active" blanks in the matrix, the key can't be pressed by the user,
+    there is no doubt as to which keys are really being pressed.
+    The ghosts will be ignored, they are KC_NO.   */
+    if (((matrix_row - 1) & matrix_row) == 0){
         return false;
-
-    // Ghost occurs when the row shares column line with other row
+    }
+    // Ghost occurs when the row shares column line with other row, blanks in the matrix don't matter
+    // If there are more than two real keys pressed and they match another row's real keys, the row will be ignored.
     for (uint8_t i=0; i < MATRIX_ROWS; i++) {
-        if (i != row && (matrix_get_row(i) & matrix_row))
+        if (i != row && __builtin_popcount((matrix_get_row(i) & matrix_ghost_check[i]) & matrix_row) > 1){
```

Comment:
The first part requires the row to have at least TWO bits set, or it returns false
```
if (((matrix_row - 1) & matrix_row) == 0){
        return false;
    }
```
so it would exit the function like this
```
if ((10 & 01) == 0){
        return false;
    }
```
now say we have:
11
01
```
if ((10 & 10) == 0){
        return false;
    }
```
now it goes to the next part
```
for (uint8_t i=0; i < MATRIX_ROWS; i++) {
    if (i != row && __builtin_popcount((matrix_get_row(i) & matrix_ghost_check[i]) & matrix_row) > 1){
        return true;
    }
}
```
```
for (uint8_t i=0; i < MATRIX_ROWS; i++) {
    if (i != row && __builtin_popcount((01) & 11) > 1){
        return true;
    }
}
```
```
for (uint8_t i=0; i < MATRIX_ROWS; i++) {
    if (i != row && __builtin_popcount(01) > 1){
        return true;
    }
}
```
```
for (uint8_t i=0; i < MATRIX_ROWS; i++) {
    if (i != row && 1 > 1){
        return true;
    }
}
```
this is false, so the next part is:

```
return false;
return false;
```
no ghost, row processed normally
I was testing some compiler settings to see if it would detect useless code like double returns, I forgot to remove the second one.